mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/MapboxTelemetry.java

```diff
@@ -0,0 +1,573 @@
+  private Boolean telemetryEnabled = null;
+
+  /**
+   * Private constructor for configuring the single instance per app.
+   */
+  private MapboxTelemetry() {
+  }
+
+  /**
+   * Primary access method (using singleton pattern)
+   *
+   * @return MapboxTelemetry
+   */
+  public static MapboxTelemetry getInstance() {
+    if (instance == null) {
+      instance = new MapboxTelemetry();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Initialize MapboxTelemetry.
+   *
+   * @param context        The context associated with the application
+   * @param accessToken    The accessToken associated with the application
+   * @param locationEngine Initialize telemetry with a custom location engine
+   */
+  public void initialize(@NonNull Context context, @NonNull String accessToken,
+                         @NonNull LocationEngine locationEngine) {
+    this.locationEngine = locationEngine;
+    initialize(context, accessToken);
+  }
+
+  /**
+   * Initialize MapboxTelemetry.
+   *
+   * @param context     The context associated with the application
+   * @param accessToken The accessToken associated with the application
+   */
+  public void initialize(@NonNull Context context, @NonNull String accessToken) {
+    if (initialized) {
+      return;
+    }
+
+    Timber.v("Initializing telemetry.");
+    this.context = context.getApplicationContext();
+    this.accessToken = accessToken;
+    if (this.context == null || TextUtils.isEmpty(this.accessToken)) {
+      throw new TelemetryException(
+        "Please, make sure you provide a valid context and access token. "
+          + "For more information, please visit https://www.mapbox.com/android-sdk.");
+    }
+
+    validateTelemetryServiceConfigured();
+    setupHttpClient();
+    checkStagingServerInformation();
+    rotateSessionId();
+    readDisplayMetrics();
+    registerBatteryUpdates();
+    loadUserPreferences();
+
+    initialized = true;
+  }
+
+  /**
+   * Checks that TelemetryService has been configured by developer
+   */
+  private void validateTelemetryServiceConfigured() {
+    try {
+      // Check the service is in the AndroidManifest.xml
+      PackageInfo packageInfo = context.getPackageManager().getPackageInfo(
+        context.getPackageName(), PackageManager.GET_SERVICES);
+      if (packageInfo.services != null) {
+        for (ServiceInfo service : packageInfo.services) {
+          if (TextUtils.equals("com.mapbox.services.android.telemetry.service.TelemetryService", service.name)) {
+            return;
+          }
+        }
+      }
+    } catch (Exception exception) {
+      Timber.w("Failed to inspect for the telemetry service: %s.", exception.getMessage());
+    }
+
+    throw new TelemetryException(
+      "Please, make sure you add the Telemetry service "
+        + "(`com.mapbox.services.android.telemetry.service.TelemetryService`) to your `AndroidManifest.xml` file. "
+        + "For more information, please visit https://www.mapbox.com/android-sdk.");
+  }
```

Comment:
Logging exception and throwing a dedicated exception can be merged together? TelemetryException wraps the other exception by passing that exception in the TelemetryException constructor?